backends-common/postgres/src/test/java/org/apache/james/backends/postgres/ConnectionThreadSafetyTest.java

```diff
@@ -0,0 +1,122 @@
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.PostgreSQLContainer;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+
+import io.r2dbc.postgresql.PostgresqlConnectionConfiguration;
+import io.r2dbc.postgresql.PostgresqlConnectionFactory;
+import io.r2dbc.postgresql.api.PostgresqlConnection;
+import io.r2dbc.postgresql.api.PostgresqlResult;
+
+@Testcontainers
+public class ConnectionThreadSafetyTest {
+    static final String DB_NAME = "james-db";
+    static final String DB_USER = "james";
+    static final String DB_PASSWORD = "1";
+    static final int NUMBER_OF_THREAD = 200;
+    static final String CREATE_TABLE_STATEMENT = "CREATE TABLE IF NOT EXISTS person (\n" +
+        "\tid serial PRIMARY KEY,\n" +
+        "\tname VARCHAR ( 50 ) UNIQUE NOT NULL\n" +
+        ");";
+
+    @Container
+    private static final GenericContainer<?> container = new PostgreSQLContainer("postgres:11.1")
+        .withDatabaseName(DB_NAME)
+        .withUsername(DB_USER)
+        .withPassword(DB_PASSWORD);
+
+    private static PostgresqlConnectionFactory connectionFactory;
+    private static SimpleJamesPostgresConnectionFactory jamesPostgresConnectionFactory;
+
+    @BeforeAll
+    static void beforeAll() {
+        connectionFactory = new PostgresqlConnectionFactory(PostgresqlConnectionConfiguration.builder()
+            .host(container.getHost())
+            .port(container.getMappedPort(5432))
+            .username(DB_USER)
+            .password(DB_PASSWORD)
+            .database(DB_NAME)
+            .build());
+        jamesPostgresConnectionFactory = new SimpleJamesPostgresConnectionFactory(connectionFactory);
+
+        connectionFactory.create().flatMapMany(connection -> connection
+            .createStatement(CREATE_TABLE_STATEMENT)
+            .execute()
+            .flatMap(PostgresqlResult::getRowsUpdated)
+            .doFinally(signalType -> connection.close())
+            .collect(Collectors.toUnmodifiableList())).blockLast();
+    }
+
+    @Test
+    void connectionShouldWorkWellWhenItIsUsedByMultipleThreads() throws Exception {
+        createData();
+
+        PostgresqlConnection connection = jamesPostgresConnectionFactory.getConnection(Domain.of("james")).block();
+
+        List<String> actual = new Vector<>();
+        ConcurrentTestRunner.builder()
+            .operation((threadNumber, step) -> {
+                connection.createStatement("SELECT id, name FROM PERSON WHERE id = $1")
+                    .bind("$1", threadNumber)
+                    .execute()
+                    .flatMap(result -> result.map((row, rowMetadata) -> row.get("id", Long.class) + "|" + row.get("name", String.class)))
+                    .doOnNext(s -> actual.add(s))
+                    .collect(Collectors.toUnmodifiableList())
+                    .block();
+            }).threadCount(NUMBER_OF_THREAD)
+            .operationCount(1)
```

Comment:
Sounds light to me as a concurrency test... 

Maybe you can lower the number of threads but they should do more operations count (which will be your step argument in the operation) . `id` could be threadNumber + step?

Still select is a read, I find that a bit light. More interested about writes. An other test with concurrent inserts maybe coupled with a select on the same operation?

Reply:
There should be no problem when a connection is used in single thread so more operations are not needed. We only need to test the connection in multi-thread env.

Added a more complicated test.

Reply:
And what about when multiple threads are trying to insert the same data?

Reply:
Maybe not necessary to go that far actually, not sure. Team thoughts?

Reply:
> And what about when multiple threads are trying to insert the same data?

It does not take much to write such tests, let's write it in my opinion. Maybe it could help us ack ahead of time some weird issues in the future.

Reply:
added insert test

Reply:
add a test of inserting same data